src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/SolutionExplorerInProcess.cs

```diff
@@ -0,0 +1,560 @@
+            if (string.IsNullOrEmpty(solutionFileFullPath))
+            {
+                throw new InvalidOperationException();
+            }
+
+            return Path.GetDirectoryName(solutionFileFullPath);
+        }
+
+        private async Task<string> GetProjectTemplatePathAsync(string projectTemplate, string languageName, CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var dte = await GetRequiredGlobalServiceAsync<SDTE, EnvDTE.DTE>(cancellationToken);
+            var solution = (EnvDTE80.Solution2)dte.Solution;
+
+            if (string.Equals(languageName, "csharp", StringComparison.OrdinalIgnoreCase)
+                && (await GetCSharpProjectTemplatesAsync(cancellationToken)).TryGetValue(projectTemplate, out var csharpProjectTemplate))
+            {
+                return solution.GetProjectTemplate(csharpProjectTemplate, languageName);
+            }
+
+            if (string.Equals(languageName, "visualbasic", StringComparison.OrdinalIgnoreCase)
+                && (await GetVisualBasicProjectTemplatesAsync(cancellationToken)).TryGetValue(projectTemplate, out var visualBasicProjectTemplate))
+            {
+                return solution.GetProjectTemplate(visualBasicProjectTemplate, languageName);
+            }
+
+            return solution.GetProjectTemplate(projectTemplate, languageName);
+        }
+
+        private async Task<ImmutableDictionary<string, string>> GetCSharpProjectTemplatesAsync(CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var hostLocale = await GetRequiredGlobalServiceAsync<SUIHostLocale, IUIHostLocale>(cancellationToken);
+            ErrorHandler.ThrowOnFailure(hostLocale.GetUILocale(out var localeID));
+
+            var builder = ImmutableDictionary.CreateBuilder<string, string>();
+            builder[WellKnownProjectTemplates.ClassLibrary] = $@"Windows\{localeID}\ClassLibrary.zip";
+            builder[WellKnownProjectTemplates.ConsoleApplication] = "Microsoft.CSharp.ConsoleApplication";
+            builder[WellKnownProjectTemplates.Website] = "EmptyWeb.zip";
+            builder[WellKnownProjectTemplates.WinFormsApplication] = "WindowsApplication.zip";
+            builder[WellKnownProjectTemplates.WpfApplication] = "WpfApplication.zip";
+            builder[WellKnownProjectTemplates.WebApplication] = "WebApplicationProject40";
+            return builder.ToImmutable();
+        }
+
+        private async Task<ImmutableDictionary<string, string>> GetVisualBasicProjectTemplatesAsync(CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var hostLocale = await GetRequiredGlobalServiceAsync<SUIHostLocale, IUIHostLocale>(cancellationToken);
```

Comment:
could the locale bit be extracted out of these methods and passed in to both the c#/vb one?  then these could be non-async which might cleanup the caller a little.

Reply:
⏱️ Yep, makes sense